### ISSUE TITLE
[Feature Request]: Add a Defined('propname') condition

### ISSUE THREAD

**Issue by ceztko** [dotnet/msbuild#9106]
### Summary

The proposal is about adding a `Defined('propname')` condition in addition to available [ones](https://learn.microsoft.com/en-us/visualstudio/msbuild/msbuild-conditions?view=vs-2022). The condition returns true if the property is present among project properties. More specifically it should evaluate to true if [`Microsoft.Build.Evaluation.Project.GetProperty(name)`](https://learn.microsoft.com/en-us/dotnet/api/microsoft.build.evaluation.project.getproperty?view=msbuild-17-netcore#microsoft-build-evaluation-project-getproperty(system-string)) returns a non null value.

### Background and Motivation

I'm trying to add preprocessor definitions and conditionally exclude files in a CSharp project, based on these definitions. A documented [approach](https://learn.microsoft.com/en-us/visualstudio/msbuild/msbuild-conditional-constructs?view=vs-2022#code-try-1), which basically suggest to to use regex directly on `$(DefineConstants)`, becomes unworkable for me even for simple use cases. 

Most build systems provides the following facilities to ease such tasks:
- Coercion of strings to boolean (eg. CMake)
- Check for definition of external properties.

MSBuild provides none of the above two. The current approach I'm using is defining externally some `FEATURE_A=1`, `FEATURE_B=1`, etc. properties (`1` is just an arbitrary value), then using these conditionals in the CSharp project:

```
  <ItemGroup Condition="$(FEATURE_A) == ''">
    <None Include="FeatureA\*.cs" />
    <Compile Remove="FeatureA\*.cs" />
  </ItemGroup>
```

### Proposed Feature

Since I'm not seeing MSBuild to implement coercion of properties to booleans, I suggest adding a `Defined('propname')` condition that will allow to implement other commonly used approaches when conditionally including/excluding stuff in msbuild projects. The example above would become:

```
  <ItemGroup Condition="!Defined('FEATURE_A')">
    <None Include="FeatureA\*.cs" />
    <Compile Remove="FeatureA\*.cs" />
  </ItemGroup>
```

**Comment by JanKrivanek** [comment]
You already mentioned the emptyness check. There are other similar techniquest that can probably be used:
* Condition="$(FEATURE_A) == ''"
* Condition="!$([System.String]::IsNullOrEmpty(`$(FEATURE_A)`))"
* [`[MSBuild]::ValueOrDefault`](https://learn.microsoft.com/en-us/visualstudio/msbuild/property-functions?view=vs-2022#msbuild-valueordefault)

But the one you mention is the most commonly used.
Can you describe scenario where that wouldn't be usable or give unexpected outcome?

**Comment by jrdodds** [comment]
> A documented [approach](https://learn.microsoft.com/en-us/visualstudio/msbuild/msbuild-conditional-constructs?view=vs-2022#code-try-1), which basically suggest to to use regex directly on `$(DefineConstants)`, becomes unworkable for me even for simple use cases.

That is an example of using a `When` element and should not be regarded as anything else. It is a terrible approach for testing `$(DefineConstants)`.

**Comment by ceztko** [comment]
> That is an example of using a `When` element and should not be regarded as anything else. It is a terrible approach for testing `$(DefineConstants)`.

I agree, but that was my effort to the effort to read the documentation. There's no good way to test  `$(DefineConstants)` and in fact I renounced attempting to do so.

**Comment by jrdodds** [comment]
> There's no good way to test `$(DefineConstants)` ...

You want to know if a certain constant is in the `$(DefineConstants)` string, correct?

The constants are separated by semi-colons, which is the default delimiter for a string of items. `$(DefineConstants)` is easy to change to an item collection. Batch on the resulting item collection to set your flags.

Roughly something like the following (but I haven't tested this code):

```xml
    <Target Name="ProcessConstants" BeforeTargets="BeforeBuild">
        <ItemGroup>
            <DefineConstants Include="$(DefineConstants)"/>
        </ItemGroup>
        <PropertyGroup>
            <FEATURE_A Condition="%(DefineConstants) == 'FEATURE_A'">true</FEATURE_A>
            <FEATURE_A Condition="$(FEATURE_A) == ''">false</FEATURE_A>

            <FEATURE_B Condition="%(DefineConstants) == 'FEATURE_B'">true</FEATURE_B>
            <FEATURE_B Condition="$(FEATURE_B) == ''">false</FEATURE_B>

            <!-- Extend with additional flag properties as needed. -->
        </PropertyGroup>
    </Target>
```

The `Condition="%(DefineConstants) == 'SOME-VALUE'"` is batching on the items.

MSBuild will coerce the string values 'true' and 'false' to Booleans. The flags are explicitly set to 'false' if empty so that a condition can be written as `Condition="$(FEATURE_A)"`.

```xml
    <Target Name="FeatureA" BeforeTargets="BeforeBuild" DependsOnTargets="ProcessConstants">
        <ItemGroup Condition="$(FEATURE_A)">
            <None Include="FeatureA\*.cs" KeepDuplicates="false" />
            <Compile Remove="FeatureA\*.cs" />
        </ItemGroup>
        <ItemGroup Condition="!$(FEATURE_A)">
            <None Remove="FeatureA\*.cs" />
            <Compile Include="FeatureA\*.cs" KeepDuplicates="false" />
        </ItemGroup>
    </Target>
```

**Comment by jrdodds** [comment]
@KalleOlaviNiemitalo Yep, it was meant to be `%(DefineConstants.Identity)`. Thanks for catching that. I edited the code example.

**Comment by ceztko** [comment]
> The Condition="%(DefineConstants) == 'SOME-VALUE'" is batching on the items.

@jrdodds Ok, this is interesting but the procedure to enable such testing is very heavy and requires a lot of boiler plate.

> MSBuild will coerce the string values 'true' and 'false' to Booleans. 

It will not if the property is not defined (null). This is done for example in CMake and it's super convenient, and I would also appreciate it in MSBuild as well (transparently or with a separate condition "operator").

Thank you a lot for trying to help in my use case, but in general the handling of null/undefined properties leaves a lot to be desired in MSBuild and I believe this goes a long way beyond the use case I initially described, so I would love to receive some feedback on the `Defined('prop')` feature I was suggesting, which I believe it would fit in MSBuild as it does in other build systems.

**Comment by jrdodds** [comment]
Instead of being limited to a `Condition` attribute, I'd like to see this implemented as an [MSBuild property function](https://learn.microsoft.com/en-us/visualstudio/msbuild/property-functions?view=vs-2022#msbuild-property-functions).

e.g.

```
$([MSBuild]::IsPropertyDefined(string propertyName))
```
